src/SMatrix.jl

```diff
@@ -123,6 +123,7 @@ end
     Base.unsafe_convert(Ptr{T}, Base.data_pointer_from_objref(m))
 end
 
+Base.promote_rule{S1,S2,T,U,L}(::Type{SMatrix{S1,S2,T,L}}, ::Type{SMatrix{S1,S2,U,L}}) = SMatrix{S1,S2,promote_type(T,U),L}
```

Comment:
`promote_rule` gets called by `promote_type`, and `promote_type` is annotated with `@pure`. So it wouldn't change anything.